File: src/main/java/com/refinedmods/refinedstorage/container/factory/CrafterManagerContainerFactory.java
package com.refinedmods.refinedstorage.container.factory;

import com.refinedmods.refinedstorage.container.CrafterManagerContainerMenu;
import com.refinedmods.refinedstorage.screen.EmptyScreenInfoProvider;
import com.refinedmods.refinedstorage.blockentity.CrafterManagerBlockEntity;
import net.minecraft.core.BlockPos;
import net.minecraft.network.FriendlyByteBuf;
import net.minecraft.world.entity.player.Inventory;
import net.minecraftforge.network.IContainerFactory;

import java.util.LinkedHashMap;
import java.util.Map;

public class CrafterManagerContainerFactory implements IContainerFactory<CrafterManagerContainerMenu> {
    @Override
    public CrafterManagerContainerMenu create(int windowId, Inventory inv, FriendlyByteBuf buf) {
        Map<String, Integer> data = new LinkedHashMap<>();

        BlockPos pos = buf.readBlockPos();

        int size = buf.readInt();

        for (int i = 0; i < size; ++i) {
            data.put(buf.readComponent().getString(), buf.readInt());
        }

        CrafterManagerContainerMenu container = new CrafterManagerContainerMenu((CrafterManagerBlockEntity) inv.player.level().getBlockEntity(pos), inv.player, windowId);

        container.setScreenInfoProvider(new EmptyScreenInfoProvider());
        container.initSlots(data);

        return container;
    }
}

File: src/main/kotlin/com/refinedmods/refinedstorage/container/factory/CrafterManagerContainerFactory.kt
package com.refinedmods.refinedstorage.container.factory

import com.refinedmods.refinedstorage.api.network.grid.IGrid.slotId
import com.refinedmods.refinedstorage.container.CrafterManagerContainer
import com.refinedmods.refinedstorage.screen.EmptyScreenInfoProvider
import com.refinedmods.refinedstorage.tile.CrafterManagerTile
import net.minecraft.entity.player.PlayerInventory
import net.minecraft.network.PacketByteBuf
import net.minecraft.util.math.BlockPos
import net.minecraftforge.fml.network.IContainerFactory
import java.util.*

class CrafterManagerContainerFactory : IContainerFactory<CrafterManagerContainer?> {
    fun create(windowId: Int, inv: PlayerInventory, buf: PacketByteBuf): CrafterManagerContainer {
        val data: MutableMap<String, Int> = LinkedHashMap()
        val pos: BlockPos = buf.readBlockPos()
        val size: Int = buf.readInt()
        for (i in 0 until size) {
            data[buf.readTextComponent().getString()] = buf.readInt()
        }
        val container = CrafterManagerContainer(inv.player.world.getBlockEntity(pos) as CrafterManagerTile?, inv.player, windowId)
        container.setScreenInfoProvider(EmptyScreenInfoProvider())
        container.initSlots(data)
        return container
    }
}